Add architectural support for shared interrupts (take 2)

LaurentiuM1234 opened the pull request:
This is an enhancement/alternative to https://github.com/zephyrproject-rtos/zephyr/pull/61331 which is meant to allow making use of shared interrupts with `IRQ_CONNECT()` (statically) and `irq_connect_dynamic` (dynamically).

A few notes:

- the shared interrupt support is transparent to the user and there's no API changes.
- I've decided to keep `shared_irq.c` in `arch/common/` for consistency reasons (it's somewhat the same thing as `sw_isr_common.c` which is placed there)
- my knowledge of linker scripts is somewhat limited but the main reason for all the linker script changes is because `_shared_irq_table` should be placed in the same memory as `_sw_isr_table` (RAM or ROM).
- **this PR is inspired from** https://github.com/zephyrproject-rtos/zephyr/pull/57146

LaurentiuM1234 committed:
arch: Add support for static shared interrupts

This commit introduces all the necessary changes for
enabling the usage of shared interrupts.

This works by using a second interrupt table: _shared_sw_isr_table
which keeps track of all of the ISR/arg pairs sharing the same
interrupt line. Whenever a second ISR/arg pair is registered
on the same interrupt line using IRQ_CONNECT(), the entry in
_sw_isr_table will be overwriten by a
(shared_isr, _shared_sw_isr_table[irq]) pair. In turn, shared_isr()
will invoke all of the ISR/arg pairs registered on the same
interrupt line.

This feature only works statically, meaning you can only make use
of shared interrupts using IRQ_CONNECT(). Attempting to dynamically
register a ISR/arg pair will overwrite the hijacked _sw_isr_table
entry.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
arch: Add support for dynamic shared interrupts

This works by overwriting z_isr_install()'s definition
(possible since the symbol is now weak) with our own definiton.

Whenever trying to register a new ISR/arg pair, z_isr_install()
will check to see if the interrupt line is already in use. If it's
not then it will not share the interrupt and will work exactly
as the old z_isr_install(), meaning it will just write the new
ISR/arg pair to _sw_isr_table.

If the interrupt line is already being used by an ISR/arg pair
then that line will become shared, meaning we'll overwrite
_sw_isr_table with our own (z_shared_isr, z_shared_sw_isr_table[irq])
pair.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
arch: Add support for dynamically disconnecting shared interrupts

This commit provides the users a way to disconnect dynamic
interrupts. This is needed because we don't want to keep
piling up ISR/arg pairs until the number of registrable
clients is reached.

This feature is only relevant for shared and dynamic interrupts.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
tests: kernel: interrupt: Add testcase for shared interrupts

This commit introduces a new testcase for shared interrupts.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented on arch/Kconfig:
```diff
@@ -403,6 +405,23 @@ config DYNAMIC_INTERRUPTS
+	default n
+	depends on GEN_SW_ISR_TABLE
+	depends on ARCH_HAS_SHARED_IRQ_SUPPORT
+	help
```
Done.

LaurentiuM1234 commented on arch/common/shared_irq.c:
```diff
@@ -0,0 +1,188 @@
+
+	/* check for out of bounds table index */
+	if (table_idx >= CONFIG_NUM_IRQS)
+		return -EINVAL;
```
Done.

LaurentiuM1234 commented on arch/common/shared_irq.c:
```diff
@@ -0,0 +1,188 @@
+
+	/* reached the client limit */
+	if (irq_data->client_num == CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS)
+		goto out_unlock;
```
Added missing braces to all `if`/`for` statements. Thanks for pointing it out!

carlocaione commented on arch/Kconfig:
```diff
@@ -403,6 +405,24 @@ config DYNAMIC_INTERRUPTS
+	depends on GEN_SW_ISR_TABLE
+	depends on ARCH_HAS_SHARED_IRQ_SUPPORT
+	select EXPERIMENTAL
+	help
```
here it is important to stress that adding support for shared IRQs is going to non-negligibly increase the image size and maybe give a ballpark estimation of that.

LaurentiuM1234 replied:
Should be a bit better. If it still doesn't stress enough the idea that enabling `SHARED_INTERRUPTS` will non-negligibly increase the image size then I'll also try to add an estimation of the increase in size like you initially suggested.

carlocaione commented on tests/kernel/interrupt/src/shared_irq.c:
```diff
@@ -0,0 +1,337 @@
+
+ZTEST(interrupt_feature, test_shared_irq)
+{
+	/* TODO: make this test work on other architectures as well.
+	 * What are the equivalents of the SGIs on the other architectures?
+	 */
```
what do you mean? `trigger_irq` should support basically all the architectures.

LaurentiuM1234 replied:
Removed comment altogether as it was misleading. It was mostly referring to which interrupt lines to use on other architectures such that the corresponding entry in `_sw_isr_table` would be `NULL/z_irq_spurious`.

carlocaione commented on tests/kernel/interrupt/src/shared_irq.c:
```diff
@@ -0,0 +1,337 @@
+	zassert_true(reset_irq_table_state_test(), "failed reset table state test");
+#endif /* CONFIG_GIC */
+
+	zassert_true(dynamic_shared_irq_test(), "failed dynamic irq test.");
```
Instead of using one single `ZTEST` and a set of `zassert_true`, use a `ZTEST` for each test.

LaurentiuM1234 replied:
This is problematic because as far as I can tell the tests are executed in alphanumerical order. This causes problems since the mini tests expect `_sw_isr_table` and `_shared_isr_table` to be in a certain state. A workaround to this problem would be to change the test naming such that a certain order is followed but IMO this is not worth the trouble.

carlocaione commented on include/zephyr/sw_isr_table.h:
```diff
@@ -61,6 +61,22 @@ struct _isr_list {
+
+struct shared_irq_data {
+	struct shared_irq_client clients[CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS];
+	uint32_t client_num;
```
```suggestion
	size_t client_num;
```

carlocaione commented on tests/kernel/interrupt/testcase.yaml:
```diff
@@ -6,3 +6,12 @@ tests:
     filter: not CONFIG_TRUSTED_EXECUTION_NONSECURE
+  arch.shared_interrupt:
+    arch_allow:
+      - arm64
+      - arm
```
I really would like to see at RISCV in this list.

LaurentiuM1234 replied:
As far as I can tell, the RISCV-based QEMU SoCs use multi-level interrupts which is currently not supported (should be added later on alongside the architectures that use it). Ideally, the support for multi-level interrupts should be added by someone a bit more familiar with it. Is this a deal-breaker?

carlocaione replied:
In that case this should be `depends on !CONFIG_MULTI_LEVEL_INTERRUPTS` but even with that you can still definitely use for example the interrupt ID 3 (software interrupt) to test this on RISCV.

LaurentiuM1234 replied:
Added `depends on !MULTI_LEVEL_INTERRUPTS` to the `SHARED_INTERRUPTS` configuration. Allowing the test to run on RISCV-based platform would mean having to remove `depends on !MULTI_LEVEL_INTERRUPTS` and adding `ARCH_HAS_SHARED_IRQ_SUPPORT` to RISCV which would mean the shared interrupts should work for all interrupt lines. This is wrong. I guess the best I can do is test this locally?

carlocaione replied:
yeah but I mean, in 90% of the RISCV platform you will have a multi level interrupt management because PLIC & company require that so I'm ready to bet that this will be never implemented for RISCV then because no one will bother with it so this will be never out of experimental and actually used.

This is why I don't like half-assessed features that work only on a very restricted selection of architectures.

Can you explain the technical difficulties in making this work also with multi level interrupts?

LaurentiuM1234 replied:
Understood. Since this is the case I'll also make it work with multi-level interrupts.


> Can you explain the technical difficulties in making this work also with multi level interrupts?

Mainly my lack of knowledge on the matter. Otherwise I expect it to be fairly easy as it's similar to how it was done for `_sw_isr_table`.

LaurentiuM1234 committed:
arch: common: sw_isr_common: Move table index computing logic to function

Since the shared IRQ code will also use the same logic to compute
the _sw_isr_table index, move the computing logic to a separate
function: z_get_sw_isr_table_idx(), which can be used by other
modules.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
**V2 changes**:

1. Removed `ARCH_HAS_SHARED_IRQ_SUPPORT` configuration. It was meant to restrict untested architectures which constituted the majority. Now's no longer the case. Also, shared interrupts are not really an architectural quirk and they should, in theory, work on all architectures.
2. Shared interrupts now work on SoC's employing multi-level interrupts.
3. Split `shared_irq.c` testcase into `static_shared_irq.c` and `dynamic_shared_irq.c` for modularity reasons.
4. Shared interrupts should now be tested on all architectures (exceptions: [1]) and are expected to work on said architectures.

[1]: there are a few architectures on which the test is not run:

1. nios2 => same as arch.interrupt tests
2. xtensa => was unable to find 2 interrupt lines that work (I only managed to get INTID 7 working). Since the test requires 2 interrupt lines this arch will be skipped until either the tests are altered such that they also work with a single interrupt (preferably should remain 2) or until a second interrupt line is found.
3. mips => was unable to get the tests working, got spurious interrupt error on any INTID in range [0, 6] (might require further investigation)
4. x86 => unmet configuration dependency for CONFIG_SHARED_INTERRUPTS.

carlocaione commented on arch/common/shared_irq.c:
```diff
@@ -0,0 +1,19 @@
+void shared_isr(const void *data)
+{
+	size_t i;
+	const struct shared_irq_data *irq_data = data;
```
any reason to not directly have
```c
void shared_isr(const struct shared_irq_data *irq_data)
```

LaurentiuM1234 replied:
Would probably require a cast each time `_sw_isr_table`'s `isr` field was assigned the value of `shared_isr`. This is not worth the trouble IMO.

carlocaione replied:
oh right, nvm.

carlocaione commented on tests/kernel/interrupt/src/dynamic_shared_irq.c:
```diff
@@ -0,0 +1,280 @@
+					  void *arg,
+					  int irq, int idx)
+{
+	int i;
```
`size_t`

carlocaione commented on tests/kernel/interrupt/src/dynamic_shared_irq.c:
```diff
@@ -0,0 +1,280 @@
+}
+
+ZTEST_SUITE(shared_irq_feature, NULL,
+	    dynamic_shared_irq_suite_setup,
+	    dynamic_shared_irq_suite_before,
+	    dynamic_shared_irq_suite_after,
```
can you add at least for one architecture a test for the statically allocated shared IRQ lines?

LaurentiuM1234 replied:
ACK. My main concern with using static shared interrupts is that the test becomes brittle. What I mean by this is that if the interrupt lines I choose are not usable on all platforms for said architecture then build errors will appear due to the number of clients exceeding the limit. Since we're limiting this to a single architecture then I guess it may not be so bad, but if more testing platforms (or even tests) are added then it may become problematic.

carlocaione replied:
yeah, but if you restrict the test to something well know like some QEMU platform or even `native_posix` you should not get into any trouble.

LaurentiuM1234 commented:
**V3 changes:**

1. The static shared interrupt test is now only executed on qemu_cortex_a53 boards. See comments in `tests/kernel/interrupt/CMakeLists.txt`
2. Excluded qemu_cortex_m0 from testing platforms. Seems like there's some failures with test_prevent_interruption w/o `CONFIG_QEMU_ICOUNT=y`. There's already plenty of ARM-based testing platforms so it should be fine.

carlocaione commented on tests/kernel/interrupt/CMakeLists.txt:
```diff
@@ -17,3 +17,14 @@ target_sources(app PRIVATE
 target_sources_ifdef(CONFIG_DYNAMIC_INTERRUPTS app PRIVATE src/dynamic_isr.c)
 target_sources_ifdef(CONFIG_X86 app PRIVATE src/regular_isr.c)
+
+if (CONFIG_SHARED_INTERRUPTS)
+# note: this test is very brittle since it requires that the chosen
+# interrupt lines are unused for all of the testing platforms. Failing
+# to meet this requirement would lead to build failures due to the number
+# of clients exceeding the limit. Still, it's important to test that the
+# static shared interrupts work properly. As such, restrict it to a single
+# platform so that we can overcome this problem.
+target_sources_ifdef(CONFIG_BOARD_QEMU_CORTEX_A53 app PRIVATE src/static_shared_irq.c)
+target_sources_ifdef(CONFIG_DYNAMIC_INTERRUPTS app PRIVATE src/dynamic_shared_irq.c)
+endif()
```
please use `platform_allow` instead of this.

LaurentiuM1234 replied:
ATM I'm under the impression that making use of `testcase.yaml` would end up becoming messier/hackier than current approach since as far as I can tell there's no way to isolate a testcase using `testcase.yaml` (please correct me if I'm wrong here). Also, `static_shared_irq.c` should only be built for `qemu_cortex_a53` and its variants otherwise we might end up having build failures as seen in previous versions of this code.  As a consequence to this, `static_shared_irq.c` would at best require a `CONFIG_GIC` instead of current `CONFIG_BOARD_QEMU_CORTEX_A53`. If the problem is that we shouldn't source files based on platform then we can quickly solve this by `#ifdef`-ing the code from `static_shared_irq.c` (not sure if this is would be better though). Will see if I can find a better, `testcase.yaml`-based solution to this.

LaurentiuM1234 replied:
Done. The test is still restricted to `qemu_cortex_a53` and its variants. Can also change it to all GIC-based SoCs like you suggested (tested locally and seems fine) but it will become more prone to errors if for whatever reason testing platforms using SGIs will be added. Either way is fine with me.

carlocaione commented on tests/kernel/interrupt/src/static_shared_irq.c:
```diff
@@ -0,0 +1,90 @@
+	irq_disable(GIC_IRQ1);
+	irq_disable(GIC_IRQ2);
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+	arch_irq_disconnect_dynamic(GIC_IRQ1, IRQ_PRIORITY, test_isr_0, 0, 0);
+	arch_irq_disconnect_dynamic(GIC_IRQ1, IRQ_PRIORITY, test_isr_1, (void *)1, 0);
+	arch_irq_disconnect_dynamic(GIC_IRQ2, IRQ_PRIORITY, test_isr_2, (void *)2, 0);
+#endif /* CONFIG_DYNAMIC_INTERRUPTS */
```
what this is for?

LaurentiuM1234 replied:
This restores the interrupt lines to their default states: `NULL/z_irq_spurious`. If the dynamic test is also ran, it will increase the possibility of `get_irq_slot()` being able to find available spots.

carlocaione replied:
This is not nice. Can you add a comment to not confuse people (like me)?

LaurentiuM1234 replied:
Done.

LaurentiuM1234 commented on tests/kernel/interrupt/CMakeLists.txt:
```diff
@@ -17,3 +17,8 @@ target_sources(app PRIVATE
+target_sources_ifdef(CONFIG_SHARED_INTERRUPTS app PRIVATE src/static_shared_irq.c)
+
+if (CONFIG_SHARED_INTERRUPTS)
+target_sources_ifdef(CONFIG_DYNAMIC_INTERRUPTS app PRIVATE src/dynamic_shared_irq.c)
```
Doesn't seem right for the static test which can be run w/o `CONFIG_DYNAMIC_INTERRUPTS=y`. Also, this would require some extra `#ifdef` logic in the dynamic test's code. I guess the "proper" way of handling this would be to split the shared IRQ test into its static and dynamic parts, each of them having a testcase in `testcase.yaml`. Apart from it being messier, it would increase the number of twister tests.

carlocaione reviewed (dismissed):
LGTM.

andyross commented on arch/common/shared_irq.c:
```diff
@@ -17,3 +18,74 @@ void shared_isr(const void *data)
+
+	/* reached the client limit */
+	if (irq_data->client_num == CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS) {
+		goto out_unlock;
```
Also, it's a silent failure.  This function doesn't have a way to report error (vs. the static variant, which will straight up fail to build on overflow!).  Might want to __ASSERT() here instead of returning a fake success, so apps have a way to catch the mistake.

LaurentiuM1234 replied:
Fixed. `__ASSERT()` is much better here than the silent approach.

andyross reviewed (changes requested):
Just noticed this new version.  Some notes, nothing awful.  I think the silently-eat-failure behavior in z_isr_install() just barely rises to a -1 for me.  Would really like to see that assert at the very least, if you don't want to modify the API to be able to fail at runtime.  The others are mostly nitpicks or suggestions.

LaurentiuM1234 commented on arch/common/shared_irq.c:
```diff
@@ -0,0 +1,19 @@
+
+	for (i = 0; i < irq_data->client_num; i++) {
+		if (irq_data->clients[i].isr) {
+			irq_data->clients[i].isr(irq_data->clients[i].arg);
```
Done.

LaurentiuM1234 commented on scripts/build/gen_isr_tables.py:
```diff
@@ -177,11 +177,34 @@ def write_address_irq_vector_table(fp, vt, nv):
+def write_shared_table(fp, shared, nv):
+    fp.write("struct shared_irq_data __shared_sw_isr_table _shared_sw_isr_table[%d] = {\n" % nv)
+    for i in range(nv):
+        if not shared[i][1]:
```
Should be better, thanks!

LaurentiuM1234 commented on arch/common/shared_irq.c:
```diff
@@ -17,3 +18,74 @@ void shared_isr(const void *data)
+		irq_data->clients[irq_data->client_num].isr =
+			_sw_isr_table[table_idx].isr;
+		irq_data->clients[irq_data->client_num].arg =
+			_sw_isr_table[table_idx].arg;
```
Fixed.

LaurentiuM1234 commented on arch/common/shared_irq.c:
```diff
@@ -17,3 +18,74 @@ void shared_isr(const void *data)
+		_sw_isr_table[table_idx].arg = irq_data;
+	}
+
+	/* don't register the same ISR/arg pair multiple times */
```
In the initial iteration of this feature `z_isr_uninstall()` didn't exist so I had to add a way to protect against piling the same `ISR/arg` pair which would happen in SOF on an i.MX platform. For some reason I've decided to keep it. Nowadays it merely provides an extra form of validation so I'm fine with removing it if need be. In the meantime, switched to using `__ASSERT()`

andyross commented on arch/common/shared_irq.c:
```diff
@@ -22,3 +23,76 @@ void shared_isr(const void *data)
+	__ASSERT(irq_data->client_num < CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS,
+		 "reached client limit");
+
+	key = k_spin_lock(&lock);
```
Not worth a -1 because it's an assert, but note that putting this before the lock introduces a race: it's possible for two (or more, technically) contexts to come through, both check that client_num has space for one more item, and then **both** grab the lock and add **two** items to the array, causing an overflow.

iuliana-prodan replied:
@LaurentiuM1234 I think that @andyross's above comment worth to be added in the code as note/comment.

LaurentiuM1234 replied:
Thanks for pointing this out. This was a slip-up on my behalf. Fixed now.

andyross reviewed (dismissed):
My sole big concern was addressed.  Looks fine here.  One note about a race condition with the error checking, but as it's an assert I'm not going to be sticky.

LaurentiuM1234 commented:
@nordicjm could you please review again?

carlocaione reviewed (changes requested):
Ups, just noticing that this PR is completely missing docs and a changelog entry. This is definitely worth a standalone section in the docs.

LaurentiuM1234 commented:
> Ups, just noticing that this PR is completely missing docs and a changelog entry. This is definitely worth a standalone section in the docs.

ACK. Would it be alright if this was added as a section in `interrupts.rst`? Also, do I need to update `release-notes-3.5.rst`?

carlocaione commented:
> ACK. Would it be alright if this was added as a section in `interrupts.rst`? Also, do I need to update `release-notes-3.5.rst`?

Yes to both questions.

LaurentiuM1234 committed:
doc: interrupts: Document support for shared interrupts

This commit adds the documentation for shared interrupts.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
doc: release: 3.5: Add note about shared interrupt support

Add a note that shared interrupts are now supported.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

dcpleung commented on include/zephyr/sw_isr_table.h:
```diff
@@ -77,6 +77,16 @@ void shared_isr(const void *data);
+ *
+ * @return corresponding index in _sw_isr_table
+ */
+unsigned int get_sw_isr_table_idx(unsigned int irq);
```
This is a public header file. This function should be prefixed by something like `z_isr_`.

LaurentiuM1234 replied:
I'm assuming this is because this function should only be used internally? Also, does the same logic apply to `shared_isr()`, which should be prefixed by a `_`?

carlescufi replied:
For now we stand by the "old" rule that non-public functions in public headers should be prefixed by `z_`. This is gonna change but it hasn't yet.

LaurentiuM1234 replied:
Thanks for the clarification ! Is there also a rule regarding the naming of structures (i.e: should `struct shared_irq_data` be turned into `struct _shared_irq_data` like in the case of `struct _isr_list`)?

dcpleung replied:
The `old` rule is to not prefix with underscore. So `z_isr_*` would be better.

LaurentiuM1234 replied:
Fixed. `get_sw_isr_table_idx` is now prefixed by a `z_`, while the shared interrupt data structures and functions are now prefixed by a `z_shared_isr` (except for the table, which has been renamed to `z_shared_sw_isr_table`)

andyross reviewed (dismissed):
Refresh +1

LaurentiuM1234 commented:
**V4 changes**:

- the static shared interrupt test is now skipped for all platforms except `qemu_cortex_a53` and friends via `#ifdef` logic.
- added documentation and changelog entry
- addressed quality of life changes suggested by @fabiobaltieri and @andyross 
- exceeding maximum number of clients and trying to register duplicate ISR/arg pairs now results in an assertion failure.
- changed naming scheme for shared interrupt data structures and functions (they are now prefixed by a `z_shared_isr` - except for the shared isr table). Also, `get_sw_isr_table_idx` is now prefixed by a `z_`.
- changed `struct shared_irq_data` to `struct z_shared_isr_table_entry` for consistency with `struct _isr_table_entry` and more accuracy.

@andyross @carlocaione @fabiobaltieri @iuliana-prodan if there's no more suggestions/issues to be addressed, could you please review one more time from scratch? There might have been a couple of changes since your initial approvals. Would like to make sure everything is fine if this get merged. Thanks!

fabiobaltieri commented on arch/common/shared_irq.c:
```diff
@@ -22,3 +23,75 @@ void z_shared_isr(const void *data)
+		shared_entry->clients[shared_entry->client_num].isr = entry->isr;
+		shared_entry->clients[shared_entry->client_num].arg = entry->arg;
+
+		shared_entry->client_num++;
```
Would not this potentially put you over CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS?

LaurentiuM1234 replied:
This can only happen if `CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS` is 1, which I'm not sure why you'd do this. Either way, will add a `BUILD_ASSERT()` to enforce the fact that `CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS` should be >= 2. Thanks for pointing this out!

andyross reviewed (approved):
Still good here

dbaluta commented:
Great work @LaurentiuM1234 !

cfriedt commented:
> - the static shared interrupt test is now skipped for all platforms except `qemu_cortex_a53` and friends via `#ifdef` logic.

So does this mean shared interrupts are only compatible with Aarch64?

I didn't get a chance to review this change before it went in, but yes, it's quite clear from the test cases added that the tests were quite tailored to arm64.

It's hard for me to tell from the tests that were added, but was any attention given to multi-level interrupts?

What are the implications in terms of size for multi-level interrupts? At a glance, it looks to be `O(k^N)` - so exponential.

LaurentiuM1234 commented:
> So does this mean shared interrupts are only compatible with Aarch64?
> 
> I didn't get a chance to review this change before it went in, but it looks like tests were tailored to arm64.

Shared interrupts work on any architecture. The restriction to `qemu_cortex_a53` is only applicable to the static test because finding 2 interrupt lines for which `_sw_isr_table` entry is NULL/`z_irq_spurious` is very hard or impossible for all testing platforms. Also, removing the restriction would make the test more prone to errors.


> It's hard for me to tell from the tests that were added, but was any attention given to multi-level interrupts?

The shared interrupt code is multi-level interrupt-aware. By this I mean it uses `z_get_sw_isr_table_idx()` to compute the table index as done by the default `z_isr_install()` (dynamic shared interrupts). For the static case, the shared interrupt table uses the same `table_idx` computed for `_sw_isr_table`.

> What are the implications in terms of size for multi-level interrupts? At a glance, it looks to be `O(k^N)` - so exponential.

I'd say the formula is something like: O(N * (sizeof(size_t) + k * sizeof(struct z_shared_isr_client))), where N is the number of interrupts and k is the value of `CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS`. Of course, to this you'd have to also factor in the size of the extra code. Since k is usually really small, I'd say it's linear. Worst case (k is comparable to N) the complexity is polynomial.
